src/dfa/client.rs

```diff
@@ -356,17 +355,9 @@ impl ClientInterface {
         }{}
         // Gather the results
         while 'condition: {
-            if self.waiting_for_received_diag.is_empty()
```

Comment:
I think this is a step in the right direction (rely on the server knowing when it is done, rather than trying to assume which group of results we should get). But I am not yet convinced that the server cannot falsely report early progress_done's.

Reply:
Can you provide more insight on what could possibly cause a falsely reported progress_end now that I've explained the pending_results flag?

Reply:
It is something I encountered a lot during early development of the DFA. The server would report progress_end and then re-report a progress_start after (which is ok behavior for the extension client, but not so much for the dfa).

Reply:
I wonder if we can introduce some test to guard against such scenario.

Reply:
Certainly possible, but:
- We would need to either pull in some sample device code into our source or check it out from the simics repository. - The test would be inconsistent, since it tries to detect a race condition of sorts.

Reply:
I can file this as an issue to create the test in another PR then.